### DISCUSSION TITLE
Ignore Java access bridge events when not focused inside the Java window

### DISCUSSION THREAD

**Discussion by thgcode** [nvaccess/nvda#14344] [Ideas]
### Is your feature request related to a problem? Please describe.
When using Insomnia to make an HTTP request, I noticed with the event tracker addon that NVDA whas receiving a lot of repetitive value change events from a scrollbar, but I was on another window.

### Describe the solution you'd like
I think we could ignore these events, however I'm opening this for discussion as I don't know if it could have negative consequences, but most of the Java access bridge event processing code makes sense when the user is inside the Java window.

### Additional context
Intellij was running a Spring project and Insomnia made an HTTP request for it.

**Comment by feerrenrut** [comment]
Given this is a discussion, I'm going to convert it to a github discussion.

@thgcode an event tracker log / output would be helpful to give us examples of the events.

**Reply by thgcode** [reply]
Here's an example. I created the following program to test.

```java
public class Main {
    public static void main(String[] args) throws Exception {
        int i = 0;
        while (true) {
            System.out.println("event " + i);
            i++;
            Thread.sleep(1000);
        }
    }
}
```

I ran this program and moved to the desktop. In this example, when I move to the Virtualbox list item on the desktop, NVDA is processing the IntelliJ's scrollbar:

EvtTracker: object: <NVDAObjects.Dynamic_SysListView32EmittingDuplicateFocusEventsListItemIAccessible object at 0x09022130>
name: Oracle VM VirtualBox
role: Role.LISTITEM
event: gainFocus
app module: <'explorer' (appName 'explorer', process ID 3580) at address 613df30>
window class name: SysListView32
IAccessible accName: 'Oracle VM VirtualBox'
IAccessibleChildID: 6
IO - speech.speech.speak (07:35:28.714) - MainThread (6980):
Speaking ['Oracle VM VirtualBox', '6 de 11', CancellableSpeech (still valid)]
DEBUG - external:globalPlugins.evtTracker.GlobalPlugin.evtDebugLogging (07:35:29.518) - MainThread (6980):
EvtTracker: object: <NVDAObjects.JAB.JAB object at 0x09022D70>
name:
role: Role.SCROLLBAR
event: valueChange
value: 3954
app module: <'appModuleHandler' (appName 'idea64', process ID 7980) at address 7592870>
window class name: SunAwtFrame
DEBUG - external:globalPlugins.evtTracker.GlobalPlugin.evtDebugLogging (07:35:29.528) - MainThread (6980):
EvtTracker: object: <NVDAObjects.JAB.JAB object at 0x090224D0>
name:
role: Role.SCROLLBAR
event: valueChange
value: 3954
app module: <'appModuleHandler' (appName 'idea64', process ID 7980) at address 7592870>
window class name: SunAwtFrame
DEBUG - external:globalPlugins.evtTracker.GlobalPlugin.evtDebugLogging (07:35:29.528) - MainThread (6980):
EvtTracker: object: <NVDAObjects.JAB.JAB object at 0x09012E70>
name:
role: Role.SCROLLBAR
event: valueChange
value: 3954
app module: <'appModuleHandler' (appName 'idea64', process ID 7980) at address 7592870>
window class name: SunAwtFrame
DEBUG - external:globalPlugins.evtTracker.GlobalPlugin.evtDebugLogging (07:35:29.528) - MainThread (6980):
EvtTracker: object: <NVDAObjects.JAB.JAB object at 0x0901F130>
name: 
role: Role.SCROLLBAR
event: valueChange
value: 3954
app module: <'appModuleHandler' (appName 'idea64', process ID 7980) at address 7592870>
window class name: SunAwtFrame
DEBUG - external:globalPlugins.evtTracker.GlobalPlugin.evtDebugLogging (07:35:30.502) - MainThread (6980):
EvtTracker: object: <NVDAObjects.JAB.JAB object at 0x07799A10>
name: 
role: Role.SCROLLBAR
event: valueChange
value: 3976
app module: <'appModuleHandler' (appName 'idea64', process ID 7980) at address 7592870>
window class name: SunAwtFrame
DEBUG - external:globalPlugins.evtTracker.GlobalPlugin.evtDebugLogging (07:35:30.502) - MainThread (6980):
EvtTracker: object: <NVDAObjects.JAB.JAB object at 0x09022CD0>
name: 
role: Role.SCROLLBAR
event: valueChange
value: 3976
app module: <'appModuleHandler' (appName 'idea64', process ID 7980) at address 7592870>
window class name: SunAwtFrame
DEBUG - external:globalPlugins.evtTracker.GlobalPlugin.evtDebugLogging (07:35:30.502) - MainThread (6980):
EvtTracker: object: <NVDAObjects.JAB.JAB object at 0x09022CF0>
name: 
role: Role.SCROLLBAR
event: valueChange
value: 3976
app module: <'appModuleHandler' (appName 'idea64', process ID 7980) at address 7592870>
window class name: SunAwtFrame
DEBUG - external:globalPlugins.evtTracker.GlobalPlugin.evtDebugLogging (07:35:30.502) - MainThread (6980):
EvtTracker: object: <NVDAObjects.JAB.JAB object at 0x0901FA90>
name: 
role: Role.SCROLLBAR
event: valueChange
value: 3976
app module: <'appModuleHandler' (appName 'idea64', process ID 7980) at address 7592870>
window class name: SunAwtFrame
DEBUG - external:globalPlugins.evtTracker.GlobalPlugin.evtDebugLogging (07:35:31.508) - MainThread (6980):
EvtTracker: object: <NVDAObjects.JAB.JAB object at 0x09022D50>
name: 
role: Role.SCROLLBAR
event: valueChange
value: 3998
app module: <'appModuleHandler' (appName 'idea64', process ID 7980) at address 7592870>
window class name: SunAwtFrame
DEBUG - external:globalPlugins.evtTracker.GlobalPlugin.evtDebugLogging (07:35:31.508) - MainThread (6980):
EvtTracker: object: <NVDAObjects.JAB.JAB object at 0x09022A90>
name: 
role: Role.SCROLLBAR
event: valueChange
value: 3998
app module: <'appModuleHandler' (appName 'idea64', process ID 7980) at address 7592870>
window class name: SunAwtFrame
DEBUG - external:globalPlugins.evtTracker.GlobalPlugin.evtDebugLogging (07:35:31.508) - MainThread (6980):
EvtTracker: object: <NVDAObjects.JAB.JAB object at 0x09012E90>
name: 
role: Role.SCROLLBAR
event: valueChange
value: 3998
app module: <'appModuleHandler' (appName 'idea64', process ID 7980) at address 7592870>
window class name: SunAwtFrame
DEBUG - external:globalPlugins.evtTracker.GlobalPlugin.evtDebugLogging (07:35:31.508) - MainThread (6980):
EvtTracker: object: <NVDAObjects.JAB.JAB object at 0x0901F230>
name: 
role: Role.SCROLLBAR
event: valueChange
value: 3998
app module: <'appModuleHandler' (appName 'idea64', process ID 7980) at address 7592870>
window class name: SunAwtFrame
IO - inputCore.InputManager.executeGesture (07:35:31.724) - winInputHook (4764):
Input: kb(laptop):NVDA+f1
DEBUG - gui.contextHelp.bindHelpEvent (07:35:31.739) - MainThread (6980):
Did context help binding for LogViewer

**Comment by CyrilleB79** [comment]
Hi @thgcode

I do not know if this scroll bar is considered a progress bar or if there is such distinction in Java Access Bridge framework.

However note that sometimes, events from background windows are needed. At least this is the case for progress bar so that NVDA may be able to report progress bars in background if the option is enabled (see Object presentation parameters).

**Reply by thgcode** [reply]
Thanks for the remainder @CyrilleB79 , Java Access Bridge has a role type for the progress bar, we'll need to make sure if we create a logic for ignoring some of the events we don't forget the progress bar... I'm not sure of the visual representation of this scrollbar specifically, but I think this event comes from the text of the Intellij's console moving.

**Comment by thgcode** [comment]
A bit unrelated, but I made this example that makes NVDA unresponsive (it becomes more apparent if you increase the delay on the sleep call), even if you move out from the Java window. In this example I hardcoded the delay, but in a real program it might delay because it is processing some data.

```java
import javax.accessibility.AccessibleContext;
import javax.swing.*;

public class Main {
    public static void main(String[] args) throws Exception {
        var frame = new JFrame();
        frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);

        JButton start = new JButton("Start");
        frame.add(start);

        JButton quit = new JButton("Quit") {
            @Override
            public AccessibleContext getAccessibleContext() {
                return new JButton.AccessibleJButton() {
                    @Override
                    public String getAccessibleName() {
                        try {
                            Thread.sleep(2000);
                        } catch (InterruptedException e) {
                            throw new RuntimeException(e);
                        }

                        return super.getAccessibleName();
                    }
                };
            }
        };

        frame.add(quit);

        frame.setVisible(true);
    }
}
```
